src/stream/src/executor/gap_fill.rs

```diff
@@ -0,0 +1,1816 @@
+                        micros,
+                    ))
+                }
+                _ => unreachable!("Time column should be Timestamp or Timestamptz"),
+            };
+            new_row_data[time_column_index] = Some(fill_time_scalar);
+            data.push(new_row_data);
+
+            fill_time = match fill_time.checked_add(*interval) {
+                Some(t) => t,
+                None => break,
+            };
+        }
+
+        // --- NEW LOGIC FOR INTERPOLATION ---
+        for (col_idx, strategy) in fill_columns {
+            if matches!(strategy, FillStrategy::Interpolate) {
+                let steps = data.len();
+                let step = Self::calculate_step(
+                    prev_row.datum_at(*col_idx).to_owned_datum(),
+                    curr_row.datum_at(*col_idx).to_owned_datum(),
+                    steps + 1,
+                );
+                if let Some(step) = step {
+                    for (i, row) in data.iter_mut().enumerate() {
+                        // The value in `row` is a copy of `prev_row`.
+                        // So we need to apply step `i+1` times.
+                        for _ in 0..=i {
+                            Self::apply_step(&mut row[*col_idx], &step);
+                        }
+                    }
+                }
+            }
+        }
+
+        for row_data in data {
+            filled_rows.push(OwnedRow::new(row_data));
+        }
+
+        Ok(filled_rows)
+    }
+}
+
+impl<S: StateStore> Execute for GapFillExecutor<S> {
+    fn execute(self: Box<Self>) -> BoxedMessageStream {
+        self.execute_inner().boxed()
+    }
+}
+
+impl<S: StateStore> GapFillExecutor<S> {
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(self: Box<Self>) {
+        let Self {
+            mut managed_state,
+            mut cache_manager,
+            schema,
+            chunk_size,
+            time_column_index,
+            fill_columns,
+            gap_interval,
+            ctx,
+            input,
+        } = *self;
+
+        let mut input = input.execute();
+
+        let barrier = expect_first_barrier(&mut input).await?;
+        let first_epoch = barrier.epoch;
+        yield Message::Barrier(barrier);
+        managed_state.init_epoch(first_epoch).await?;
+
+        // Start with an empty cache - use lazy loading strategy.
+        // Data will be loaded on-demand when gap filling operations require it.
+
+        // Calculate and validate gap interval once at initialization
+        let dummy_row = OwnedRow::new(vec![]);
+        let interval_datum = gap_interval.eval_row_infallible(&dummy_row).await;
+        let interval = interval_datum
+            .ok_or_else(|| anyhow::anyhow!("Gap interval expression returned null"))?
+            .into_interval();
+
+        // Validate that gap interval is not zero
+        if interval.months() == 0 && interval.days() == 0 && interval.usecs() == 0 {
```

Comment:
Looks good to be defensive here. We can also check this earlier during binding in the frontend.